Edit \data page to display 'Hello Malachi'

malachibre opened the pull request.

sehtia-raghav commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +25,35 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  public ArrayList<String> randomFacts;
```
Can we make randomFacts private? 

Aside from intentionally public APIs, its recommended practice that all remaining classes and members should use the lowest visibility possible.

malachibre replied:
In a later step it actually asks us to remove the stuff that we did in this step, so I don't have a randomFacts variable anymore.

sehtia-raghav commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +25,35 @@
+    response.getWriter().println(json);
+  }
+
+  public String convertToJson(ArrayList<String> randomFacts) {
```
Would gson work here to convert a list to JSON? I think it creates a JSON array fyi
See https://sites.google.com/site/gson/gson-user-guide#TOC-Collections-Examples

malachibre replied:
This is also something that we removed in a later step. But thanks, I was originally confused on how to use Gson to create the key-value pairs.

sehtia-raghav commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +25,35 @@
+    response.getWriter().println(json);
+  }
+
+  public String convertToJson(ArrayList<String> randomFacts) {
```
Since randomFacts is also the name of a class variable, it might help readability to name this parameter something else.

Although the method is referring to the randomFacts param rather than the randomFacts class var, when reading the method it could get confusing which is being referred to.

malachibre replied:
I changed the name of randomFacts to comments, and made that variable private and final.

malachibre replied:
Done.

sehtia-raghav commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +25,35 @@
+    response.getWriter().println(json);
+  }
+
+  public String convertToJson(ArrayList<String> randomFacts) {
```
Does the convertToJson method need to be public? If not let's make it private to restrict visibility as much as we can.

malachibre replied:
Done.

malachibre commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +25,35 @@
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    init();
```
Done.

malachibre replied:
I removed the init() method.

malachibre commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +25,35 @@
+  }
+
+  public String convertToJson(ArrayList<String> randomFacts) {
+      String json = "{";
+      json += "\"greeting\": ";
+      json += "\"" + randomFacts.get(0) + "\"";
+      json += ", ";
+      json += "\"favoriteColor\": ";
+      json += "\"" + randomFacts.get(1) + "\"";
+      json += ", ";
+      json += "\"major\": ";
+      json += "\"" + randomFacts.get(2) + "\"";
+      json += "}";
+      return json;
```
Alright, I imported Gson.

malachibre commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +25,35 @@
+
+  @Override
+  public void init() {
+      randomFacts = new ArrayList<String>();
```
Done.

malachibre commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +25,35 @@
+  }
+
+  public String convertToJson(ArrayList<String> randomFacts) {
+      String json = "{";
```
Done.

malachibre commented on portfolio/src/main/webapp/index.html:
```diff
@@ -61,6 +63,8 @@ <h1 id="random-facts">Random Facts</h1>
    Once when I was on a skateboard and once when I was on a bike.
-   One thing that others point out is how I can remember the tiniest things.
+   One thing that others point out is how I can remember the tiniest details about things.
+<br><br>
```
I ended up not even needing the <br> tags.

malachibre commented on portfolio/src/main/webapp/index.html:
```diff
@@ -61,6 +63,8 @@ <h1 id="random-facts">Random Facts</h1>
-   One thing that others point out is how I can remember the tiniest things.
+   One thing that others point out is how I can remember the tiniest details about things.
+<br><br>
+<button id="get-facts" onclick="getFacts()"> Pull Facts 
```
Done.

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
+/** Retrieves data from the /data page and displays it. */
+function getGreeting() {
+  fetch('/data').then((response) => response.json()).then((json) => {
+    document.getElementById('greeting').innerHTML = json.greeting;
```
Okay, I was wondering if there was a difference between the two.

malachibre commented on portfolio/src/main/webapp/style.css:
```diff
@@ -4,6 +4,11 @@ body {
 }
 
+#get-greeting {
+    height: 5%;
```
Done.

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
+  });
+}
+
+/** Retreives data from /data and passes it into the 
+ *  createPElements function
+ */
```
Done.

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
+  });
+}
+
+/** Retreives data from /data and passes it into the 
```
English is hard haha.

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
+ *  createPElements function
+ */
+function getFacts() {
+  fetch('/data').then((response) => response.json()).then((json) => {
+    createPElements(json);
+  });
```
Done.

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
+  });
+}
+
+/** Creates p elements for the text favoriteColor and majors 
+ *  attrtibutes from json 
  */
```
Done.

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
-// Cycles through 3 different pictures of myself.
 let pictureNumber = 0;
+
+/**  Cycles through 3 different pictures of myself. */
```
Done.

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
+ *  attrtibutes from json 
  */
+function createPElements(json) {
+  
```
Done.

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
+  const majorText = document.createElement("p");
+  majorText.innerText = json.major;  
+
+  document.body.appendChild(favoriteColorText);
+  document.body.appendChild(majorText);
```
Yea I had just though about that issue. I was gonna change the button to a toggle button.

malachibre commented on portfolio/src/main/webapp/index.html:
```diff
@@ -61,6 +63,8 @@ <h1 id="random-facts">Random Facts</h1>
-   One thing that others point out is how I can remember the tiniest things.
+   One thing that others point out is how I can remember the tiniest details about things.
+<br><br>
+<button id="get-facts" onclick="getFacts()"> Pull Facts 
```
Done.

malachibre replied:
Done.

malachibre replied:
I deleted the getGreeting() and getFacts() functions, but now I have the getComments() function which needs to be ran every time the page loads. Right now I have <body onLoad="getComments()"> defined in the html, but the styleguide says to stay away from optional tags, so should I find a way to put that in my JS?

malachibre commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,45 @@
+}
+
+/** Creates p elements for the text favoriteColor and majors 
+ *  attrtibutes from json 
```
Done.

malachibre commented:
It's weird because in step 4 of the week-3 walk-through, it tells us to remove the hard-coded stuff that we did in step 3. So that means I removed getGreeting(), getFacts() in JavaScript, the randomFacts variable in Java, and I changed them to focus on the comments. I ended up integrating the suggestions you two made here with the changes I made in step 4.

sehtia-raghav commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,27 @@
+  });
+}
+
+/** Creates p elements and sets the text to comments pulled from \data page*/
```
nit: I think we can use the full-name here so it's clear that p means paragraph.

I.e. "Creates paragraph elements"

sehtia-raghav commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,27 @@
+  });
+}
+
+/** Creates p elements and sets the text to comments pulled from \data page*/
```
nit: Let's add a period at end of statement and trailing space before close comment

malachibre replied:
Done.

sehtia-raghav commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,27 @@
   headerSelfie.src = `images/${imageNames[pictureNumber++ % imageNames.length]}.jpg`;
 }
+
+
```
nit: I think we could remove one of the two newlines at the end of the file?

corycook commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,27 @@
   const imageNames = ["campus", "suit", "main-selfie"];
   headerSelfie.src = `images/${imageNames[pictureNumber++ % imageNames.length]}.jpg`;
-}
+}
```
Nit: have 1 empty line at the end of the file

malachibre replied:
Done.

corycook commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,27 @@
+function createPElement(commentText) {
+  const commentElement = document.createElement("p");
+  commentElement.innerText = commentText
+  document.body.appendChild(commentElement);
```
A container element should be used to hold the comments. This will allow the comments to be styled and positioned on the page with CSS.

malachibre replied:
Okay, I added a div box.

corycook commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,27 @@
+/** Creates paragraph elements and sets the text to comments 
+ *   pulled from \data page. 
  */
+function createPElement(commentText) {
```
This also adds the element to the page. The name should probably reflect this. Something like `displayComment` might be better.

malachibre replied:
Done.

corycook commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -14,7 +14,9 @@
 
 package com.google.sps.servlets;
 
+import com.google.gson.Gson;
```
In order to use Gson you have to add it to your maven package dependencies in **pom.xml**. 

See https://github.com/google/gson/blob/master/UserGuide.md#using-gson-with-maven

Add this to your **pom.xml** file inside of `<dependencies> ... </dependencies>`
```xml
<!--  Gson: Java to Json conversion -->
<dependency>
  <groupId>com.google.code.gson</groupId>
  <artifactId>gson</artifactId>
  <version>2.8.6</version>
  <scope>compile</scope>
</dependency>
```

malachibre replied:
Ah I added it in, but I forgot to stage my pom changes.

sehtia-raghav commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,14 +12,27 @@
+}
+
+/** Creates paragraph elements and sets the text to comments 
+ *   pulled from \data page. 
```
If there are multiple lines in a javadoc it's practice to keep the comment-markers separate (without text).
Ie.
/**
 * some text here
 * some more tex
 */

See https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line

malachibre replied:
Done.

malachibre commented on portfolio/src/main/webapp/index.html:
```diff
@@ -2,6 +2,7 @@
 <meta charset="utf-8">
 <title>My Portfolio</title>
 <link rel="stylesheet" href="style.css">
+<body onload="getComments()">
```
Done.

malachibre commented on portfolio/src/main/webapp/index.html:
```diff
@@ -61,6 +62,11 @@ <h1 id="random-facts">Random Facts</h1>
-   One thing that others point out is how I can remember the tiniest things.
+   One thing that others point out is how I can remember the tiniest details about things.
+<form action="/data" method="POST">
+  <p>Add a comment!<br>
```
Done.

malachibre commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +26,20 @@
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
```
Done.

malachibre commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -24,9 +26,20 @@
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    response.setContentType("application/json");
```
Done.

corycook commented on portfolio/src/main/webapp/script.js:
```diff
@@ -12,12 +12,28 @@
+  });
+}
+
+window.addEventListener('DOMContentLoaded', getComments, false);
```
Nit: move this line to the bottom of the file, top of the file, or at least below `displayComment`.

`getComments` and `displayComments` are closely related (one calls the other) and should be next to each other in the file.

malachibre replied:
Done.

corycook commented on portfolio/src/main/webapp/style.css:
```diff
@@ -86,3 +86,7 @@ h3 {
 }
+
+form {
+    padding: 0px;
```
Nit: Use 2 spaces for indent

See https://google.github.io/styleguide/htmlcssguide.html#Indentation

malachibre replied:
Done.

corycook commented on portfolio/src/main/webapp/index.html:
```diff
@@ -61,6 +61,12 @@ <h1 id="random-facts">Random Facts</h1>
    Once when I was on a skateboard and once when I was on a bike.
-   One thing that others point out is how I can remember the tiniest things.
+   One thing that others point out is how I can remember the tiniest details about things.
+<form action="/data" method="POST">
```
`method="POST"`

Use lowercase for attribute values.

See https://google.github.io/styleguide/htmlcssguide.html#Capitalization

```html
<form action="/data" method="post">
```

malachibre replied:
Done.

corycook commented on portfolio/src/main/webapp/index.html:
```diff
@@ -8,14 +8,14 @@
   <div id="greeting-container">
-    <h1>Welcome to my portfolio</h1>
+    <h1>Malachi's portfolio</h1>
     <h3>My name is Malachi Brewer.
```
These should really be `<p>` instead of `<h3>`. If you want to change the font size you can do so with CSS

```
#greeting-container p {
  font-size: 2rem; /** double base font size */
}
```

malachibre replied:
Done.

malachibre commented on portfolio/src/main/webapp/index.html:
```diff
@@ -61,6 +61,12 @@ <h1 id="random-facts">Random Facts</h1>
+<form action="/data" method="POST">
+  <p>Add a comment!
+  <input type="text" name="comment" value="comment">
+  <input type="submit"/>
```
Done.
